templates/js-langchain/src/vector_index_cache.js

```diff
@@ -0,0 +1,51 @@
+import { Actor } from 'apify';
+import { createHash } from 'crypto';
+import { readFile } from 'node:fs/promises';
+import tempWrite from 'temp-write';
+import tar from 'tar';
+
+const VECTOR_INDEX_CACHE_STORE_NAME = 'vector-index-cache';
+
+/**
+ * Generates vector index cache key as a hash of provided configuration object.
+ *
+ * @param {Object} config
+ * @returns {String}
+ */
+function getIndexCacheKey(config) {
+    const hash = createHash('md5').update(JSON.stringify(config)).digest('hex');
+
+    return `${hash}.tar`;
+}
+
+/**
+ * Caches a vector index from `indexPath` in a VECTOR_INDEX_CACHE_STORE_NAME key-value store a key generated from `config`.
+ *
+ * @param {Object} config
+ * @param {String} indexPath
+ */
+export async function cacheVectorIndex(config, indexPath) {
+    const vectorIndexCacheStore = await Actor.openKeyValueStore(VECTOR_INDEX_CACHE_STORE_NAME);
+    const tempFilePath = await tempWrite(tar.c({}, [indexPath]));
+    const gzipedVectorIndex = await readFile(tempFilePath);
+
+    await vectorIndexCacheStore.setValue(getIndexCacheKey(config), gzipedVectorIndex, { contentType: 'application/tar' });
+}
+
+/**
+ * Fetches a cached vector index from a VECTOR_INDEX_CACHE_STORE_NAME key-value store and extracts it to the current directory.
+ *
+ * @param {Object} config
+ * @returns {Boolean} noting if the vector index was found in the cache
+ */
+export async function retrieveVectorIndex(config) {
+    const vectorIndexCacheStore = await Actor.openKeyValueStore(VECTOR_INDEX_CACHE_STORE_NAME);
+
+    const vectorIndexRecord = await vectorIndexCacheStore.getValue(getIndexCacheKey(config));
+    if (!vectorIndexRecord) return false;
+
+    const tempFilePath = await tempWrite(vectorIndexRecord);
+    await tar.x({ file: tempFilePath, strip: 1, C: '.' });
```

Comment:
You don't need the temporary file here, you could just do something like `await stream.finished(Readable.from(vectorIndexRecord).pipe(tar.x(...)))`

Reply:
And the writing you could do with something like https://www.npmjs.com/package/concat-stream instead of the temporary files

Reply:
This is the thing. I originally went for this approach, but I thought the Apify SDK (`getValue`/`setValue`) does not have streaming support. But it does, you just need to tell the content type ahead 🤦 so everything is like 10x simpler now.

You don't even need any extra package, just core ones:

```
import { finished } from 'node:stream/promises';

await finished(Readable.from(vectorIndexRecord).pipe(tar.x({ strip: 1, C: '.' })));
```

```
const packedVectorIndexStream = tar.c({}, [indexPath]);

await vectorIndexCacheStore.setValue(getIndexCacheKey(config), packedVectorIndexStream, { contentType: 'application/tar' });
```